pkg/storage/driver/memory.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package driver // import "k8s.io/helm/pkg/storage/driver"
+
+import (
+	"sync"
+
+	rspb "k8s.io/helm/pkg/proto/hapi/release"
+)
+
+// Memory is an in-memory storage driver implementation.
+type Memory struct {
+	sync.RWMutex
+	cache map[string]*rspb.Release
+}
+
+func NewMemory() *Memory {
+	return &Memory{cache: map[string]*rspb.Release{}}
+}
+
+// Get returns the release named by key.
+func (mem *Memory) Get(key string) (*rspb.Release, error) {
+	defer unlock(mem.rlock())
+
+	if rls, ok := mem.cache[key]; ok {
+		return rls, nil
+	}
+	return nil, ErrReleaseNotFound
+}
+
+// All returns all releases whose status is not Status_DELETED.
```

Comment:
Given the way that this API is evolving, I'm wondering if `All()` should return all releases, regardless of the Status flag, and let a higher-level API remove the deleted ones.

As it is, there is effectively no way to list deleted items with this API, which will cause problems when implementing rollback or when adding flags like in #973.

Maybe just allowing a boolean flag (`All(deleted bool)`) would solve this problem.

Part of the issue is that we have conflated two different notions of `delete` now. There's the delete on the driver, which indicates whether the data item is to be removed from storage (a hopefully infrequent action), and the `Status_DELETED`, which indicates whether the resources associated with a release have been deleted from Kubernetes. Maybe we should just make it so that the driver doesn't necessarily know anything about `Status_DELETED`.